README.md
## ciro-maciel - webpack-config [![deploy to Prod](https://github.com/ciro-maciel/webpack-config/actions/workflows/toProd.yml/badge.svg)](https://github.com/ciro-maciel/webpack-config/actions/workflows/toProd.yml) ![npm](https://img.shields.io/npm/v/@ciro-maciel/webpack-config) ![NPM](https://img.shields.io/npm/l/@ciro-maciel/webpack-config)

webpack configuration for development Applications

This is our shared [Webpack](http://webpack.github.io) config used for front-end projects at ciro-maciel.me. It compiles JavaScript with [Babel](https://babeljs.io). It is also configured to add hashes to filenames for easy caching.

### Getting Started

Install this package via NPM:

```
npm install @ciro-maciel/webpack-config --save-dev
```

Add some scripts to your `package.json`:

```js
{
  // ...
  "scripts": {
    "start": "webpack --env=dev",
    "build": "webpack --env=prod",
  }
}
```

Create a `webpack.config.js` in your project directory, and set it up like so:

```js
const configure = require("@ciro-maciel/webpack-config");

module.exports = configure({
  entry: {
    // Add your bundles here, so in this case
    // ./src/app.js ==> ./dist/app-[hash].js
    app: "./src/app.js",
  },
  __dirname,
  // Override any other Webpack settings here!
  // see: https://webpack.js.org/configuration/
});
```

Now you can run `npm start` to build with source maps and watch for changes, and `npm run build` to build optimized assets for production! If you need to further customize your build, you can pass any overrides in to the configure function.

![base project performance](https://github.com/ciro-maciel/webpack-config/blob/main/docs/Screen%20Shot%202021-03-13%20at%2007.52.22%402x.jpg)

### License

&copy; ciro-maciel.me. Our Webpack config is free software, and may be redistributed under the terms specified in the [LICENSE](https://github.com/ciro-maciel/webpack-config/blob/main/LICENSE) file. The name and logo for ciro-maciel.me are trademarks of Ciro Cesar Maciel and may not be used without permission.


CHANGELOG.md
## [1.10.3](https://github.com/ciro-maciel/webpack-config/compare/v1.10.2...v1.10.3) (2021-03-12)


### Bug Fixes

* CleanWebpackPlugin ([bc5449d](https://github.com/ciro-maciel/webpack-config/commit/bc5449d7e15ebd1b2c15f43288a1bba5968faf69))

## [1.10.2](https://github.com/ciro-maciel/webpack-config/compare/v1.10.1...v1.10.2) (2021-03-12)


### Bug Fixes

* WorkboxPlugin ([0829074](https://github.com/ciro-maciel/webpack-config/commit/0829074eae7e47f7a705699155ea855607da8e9a))

## [1.10.1](https://github.com/ciro-maciel/webpack-config/compare/v1.10.0...v1.10.1) (2021-03-12)


### Bug Fixes

* modifyURLPrefix ([545ce46](https://github.com/ciro-maciel/webpack-config/commit/545ce46548df014612f2805d506bbf9e002b134c))

# [1.10.0](https://github.com/ciro-maciel/webpack-config/compare/v1.9.1...v1.10.0) (2021-03-12)


### Features

* **worker:** add WorkboxPlugin ([60bbb2f](https://github.com/ciro-maciel/webpack-config/commit/60bbb2f3db9919a7e7980e830ff47809dc07e291))

## [1.9.1](https://github.com/ciro-maciel/webpack-config/compare/v1.9.0...v1.9.1) (2021-03-12)


### Bug Fixes

* devtool ([64ad62d](https://github.com/ciro-maciel/webpack-config/commit/64ad62d216a38ccbc9eefad7f7a25e37061e6d72))

# [1.9.0](https://github.com/ciro-maciel/webpack-config/compare/v1.8.1...v1.9.0) (2021-03-12)


### Features

* markdown-loader ([7a3d06d](https://github.com/ciro-maciel/webpack-config/commit/7a3d06da9956d242670f98599f6997696207b9be))

## [1.8.1](https://github.com/ciro-maciel/webpack-config/compare/v1.8.0...v1.8.1) (2021-03-11)


### Bug Fixes

* sourceMaps ([992f42c](https://github.com/ciro-maciel/webpack-config/commit/992f42cb86c972cadf829b8c9fd0e1fd20c2399a))

# [1.8.0](https://github.com/ciro-maciel/webpack-config/compare/v1.7.1...v1.8.0) (2021-03-11)


### Features

* TerserPlugin ([7c8807f](https://github.com/ciro-maciel/webpack-config/commit/7c8807f0848565c75b245eb46e7fff44fa7f1a38))

## [1.7.1](https://github.com/ciro-maciel/webpack-config/compare/v1.7.0...v1.7.1) (2021-03-11)


### Bug Fixes

* css performance ([68a2f4d](https://github.com/ciro-maciel/webpack-config/commit/68a2f4d8bf00f169ca10f6e4a82433097ba5a93f))

# [1.7.0](https://github.com/ciro-maciel/webpack-config/compare/v1.6.3...v1.7.0) (2021-03-11)


### Features

* CssMinimizerPlugin ([dbf81fb](https://github.com/ciro-maciel/webpack-config/commit/dbf81fbb69c5d497fcd6febeaf313a225d37a6e9))

## [1.6.3](https://github.com/ciro-maciel/webpack-config/compare/v1.6.2...v1.6.3) (2021-03-11)


### Bug Fixes

* asset/resource ([ac1dd7d](https://github.com/ciro-maciel/webpack-config/commit/ac1dd7dc03fbee71764d5b6c6be485d9077b5df3))

## [1.6.2](https://github.com/ciro-maciel/webpack-config/compare/v1.6.1...v1.6.2) (2021-03-11)


### Bug Fixes

* publicPath ([61c2a27](https://github.com/ciro-maciel/webpack-config/commit/61c2a27f2f152dc4dc0d4ef9c5448f647b230ebe))

## [1.6.1](https://github.com/ciro-maciel/webpack-config/compare/v1.6.0...v1.6.1) (2021-03-11)


### Bug Fixes

* FaviconsWebpackPlugin ([ad2dbd0](https://github.com/ciro-maciel/webpack-config/commit/ad2dbd09d406537ffa95e1c3c30f3224fb1df52c))

# [1.6.0](https://github.com/ciro-maciel/webpack-config/compare/v1.5.1...v1.6.0) (2021-03-11)


### Features

* FaviconsWebpackPlugin ([6f09281](https://github.com/ciro-maciel/webpack-config/commit/6f09281bab1b2ac064386f3e5b3025049d3b3276))

## [1.5.1](https://github.com/ciro-maciel/webpack-config/compare/v1.5.0...v1.5.1) (2021-03-10)


### Bug Fixes

* images path ([3863025](https://github.com/ciro-maciel/webpack-config/commit/386302547f7ea9c9572abd2716f9f63f4a21a79e))

# [1.5.0](https://github.com/ciro-maciel/webpack-config/compare/v1.4.1...v1.5.0) (2021-03-10)


### Features

* **asset:** using webpack ([d114cf7](https://github.com/ciro-maciel/webpack-config/commit/d114cf76d7511e5305539ae822d709c582e235a0))

## [1.4.1](https://github.com/ciro-maciel/webpack-config/compare/v1.4.0...v1.4.1) (2021-03-10)


### Bug Fixes

* resource ([3486b7b](https://github.com/ciro-maciel/webpack-config/commit/3486b7b118e46b5f398e721346be4af4fb9172cd))

# [1.4.0](https://github.com/ciro-maciel/webpack-config/compare/v1.3.11...v1.4.0) (2021-03-10)


### Features

* asset/resource ([b869432](https://github.com/ciro-maciel/webpack-config/commit/b869432d5181d3a5cae89e09c411c2b5c667164f))

## [1.3.11](https://github.com/ciro-maciel/webpack-config/compare/v1.3.10...v1.3.11) (2021-03-10)


### Bug Fixes

* path ([d347792](https://github.com/ciro-maciel/webpack-config/commit/d347792a9018061d7f744cf2a26434d2070aaf79))

## [1.3.10](https://github.com/ciro-maciel/webpack-config/compare/v1.3.9...v1.3.10) (2021-03-10)


### Bug Fixes

* publicPath ([9c81337](https://github.com/ciro-maciel/webpack-config/commit/9c813374edbdb468ced23a291998257cf3bc84aa))

## [1.3.9](https://github.com/ciro-maciel/webpack-config/compare/v1.3.8...v1.3.9) (2021-03-10)


### Bug Fixes

* filename: path.resolve ([c339fc3](https://github.com/ciro-maciel/webpack-config/commit/c339fc314b546ad3f94905758b512cf7fadde8d8))

## [1.3.8](https://github.com/ciro-maciel/webpack-config/compare/v1.3.7...v1.3.8) (2021-03-10)


### Bug Fixes

* filename ([873eb4c](https://github.com/ciro-maciel/webpack-config/commit/873eb4cb2d41cae5025922eb5aa870b0042ff817))

## [1.3.7](https://github.com/ciro-maciel/webpack-config/compare/v1.3.6...v1.3.7) (2021-03-10)


### Bug Fixes

* path ([11e8fd5](https://github.com/ciro-maciel/webpack-config/commit/11e8fd57ed13c71534d6c5e6b1c4f9083c7caa04))

## [1.3.6](https://github.com/ciro-maciel/webpack-config/compare/v1.3.5...v1.3.6) (2021-03-10)


### Bug Fixes

* publicPath ([9f3b115](https://github.com/ciro-maciel/webpack-config/commit/9f3b11565c8bb087af314a064f82e05c181e2571))

## [1.3.5](https://github.com/ciro-maciel/webpack-config/compare/v1.3.4...v1.3.5) (2021-03-10)


### Bug Fixes

* less ([3510ae0](https://github.com/ciro-maciel/webpack-config/commit/3510ae041e016e7f154b7f3a05f8b35f56652c7e))


main.js
const path = require("path");
const webpack = require("webpack");
const Dotenv = require("dotenv-webpack");
const { merge } = require("webpack-merge");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const ProgressBarPlugin = require("progress-bar-webpack-plugin");

const ImageMinimizerPlugin = require("image-minimizer-webpack-plugin");

const CopyPlugin = require("copy-webpack-plugin");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const CompressionPlugin = require("compression-webpack-plugin");
const { BundleAnalyzerPlugin } = require("webpack-bundle-analyzer");

const baseConfig = {
  // https://webpack.js.org/configuration/devtool/
  devtool: "source-map",
  entry: "./src/index.js",
  output: {
    path: path.resolve(__dirname, "../www", "assets/js/"),
    filename: "[contenthash:12].js",
    chunkFilename: "[chunkhash:12].js",
  },
  performance: { hints: false },
  optimization: {
    // https://webpack.js.org/plugins/split-chunks-plugin/
    splitChunks: {
      chunks: "all",
      maxSize: 600000,
      minChunks: 5,
      cacheGroups: {
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          priority: -10,
        },
        default: {
          minChunks: 2,
          priority: -20,
          reuseExistingChunk: true,
        },
      },
    },
  },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        loader: "babel-loader",
        options: {
          babelrc: false,
          presets: ["@babel/preset-env", "@babel/preset-react"],
        },
      },
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"],
      },
      {
        test: /\.(woff|woff2)(\?v=\d+\.\d+\.\d+)?$/,
        use: {
          loader: "url-loader",
          options: {
            limit: 50000,
          },
        },
      },
      {
        test: /\.(png|jp(e*)g|svg|md)$/,
        use: [
          {
            loader: "url-loader",
          },
          {
            loader: ImageMinimizerPlugin.loader,
            options: {
              severityError: "warning",
              minimizerOptions: {
                plugins: [
                  ["optipng", { optimizationLevel: 5 }],
                  [
                    "svgo",
                    {
                      plugins: [
                        {
                          removeViewBox: false,
                        },
                      ],
                    },
                  ],
                ],
              },
            },
          },
        ],
      },
    ],
  },
  plugins: [
    new webpack.ContextReplacementPlugin(/moment[\/\\]locale$/, /pt-br/),
    new ProgressBarPlugin({
      format: "Build [:bar] :percent (:elapsed seconds)",
      clear: false,
    }),
  ],
  // https://webpack.js.org/configuration/resolve/#resolvealias
  resolve: {
    alias: {
      components: path.resolve(__dirname, "../src/components/index.js"),
      containers: path.resolve(__dirname, "../src/containers/index.js"),
      providers: path.resolve(__dirname, "../src/providers/index.js"),
      hooks: path.resolve(__dirname, "../src/hooks/index.js"),
      utils: path.resolve(__dirname, "../src/utils/index.js"),
    },
  },
};

const prodConfig = {
  mode: "production",
  output: {
    publicPath: "assets/js/",
  },
  plugins: [
    new CleanWebpackPlugin({
      verbose: true,
      cleanOnceBeforeBuildPatterns: [
        path.join(__dirname, "../www/assets/**/*"),
      ],
    }),
    new HtmlWebpackPlugin({
      template: "src/index.html",
      filename: "../../index.html",
      showErrors: true,
      chunksSortMode: "auto",
      minify: {
        removeComments: false,
        collapseWhitespace: true,
        useShortDoctype: true,
        minifyJS: true,
        minifyCSS: true,
        minifyURLs: true,
      },
    }),
    new Dotenv({
      path: path.join(__dirname, `/.env.prod`),
      safe: true,
      systemvars: true,
    }),
    new CopyPlugin({
      patterns: [{ from: "src/assets/img/", to: "../img/" }],
    }),
    new CompressionPlugin({
      test: /\.js$/,
    }),
    new BundleAnalyzerPlugin({
      analyzerMode: "static",
      openAnalyzer: false,
      reportFilename: "../../analyzer.html",
    }),
  ],
};

const devConfig = {
  mode: "development",
  output: {
    publicPath: "/",
  },
  plugins: [
    new HtmlWebpackPlugin({
      template: "src/index.html",
      filename: "index.html",
      showErrors: true,
      chunksSortMode: "auto",
      minify: {
        removeComments: false,
        collapseWhitespace: true,
        useShortDoctype: true,
        minifyJS: true,
        minifyCSS: true,
        minifyURLs: true,
      },
    }),
    new Dotenv({
      path: path.join(__dirname, `/.env.dev`),
      safe: true,
      systemvars: true,
    }),
  ],
  devServer: {
    contentBase: path.join(__dirname, "../../../www/"),
    index: "index.html",
    compress: true,
    open: true,
    port: 9090,
    hot: true,
    historyApiFallback: true,
  },
};

module.exports = (options = {}) => (env) =>
  merge(baseConfig, env.dev ? devConfig : prodConfig, options);


index.js
const path = require("path");
const webpack = require("webpack");
const Dotenv = require("dotenv-webpack");
const { merge } = require("webpack-merge");
const TerserPlugin = require("terser-webpack-plugin");
const WorkboxPlugin = require("workbox-webpack-plugin");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const ProgressBarPlugin = require("progress-bar-webpack-plugin");
const FaviconsWebpackPlugin = require("favicons-webpack-plugin");
const CssMinimizerPlugin = require("css-minimizer-webpack-plugin");
const { BundleAnalyzerPlugin } = require("webpack-bundle-analyzer");

const baseConfig = (dirPath) => ({
  // https://webpack.js.org/configuration/devtool/
  devtool: "source-map",
  entry: "./src/index.js",
  output: {
    path: path.resolve(dirPath, "../www", "assets/"),
    filename: "js/[contenthash:12].js",
    chunkFilename: "js/[chunkhash:12].js",
  },
  performance: { hints: false },
  optimization: {
    // https://webpack.js.org/plugins/split-chunks-plugin/
    splitChunks: {
      chunks: "all",
      maxSize: 600000,
      minChunks: 5,
      cacheGroups: {
        vendors: {
          test: /[\\/]node_modules[\\/]/,
          priority: -10,
        },
        default: {
          minChunks: 2,
          priority: -20,
          reuseExistingChunk: true,
        },
      },
    },
  },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        loader: "babel-loader",
        options: {
          babelrc: false,
          presets: ["@babel/preset-env", "@babel/preset-react"],
          plugins: [
            [
              "import",
              { libraryName: "antd", libraryDirectory: "es", style: true },
            ],
          ],
        },
      },
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"],
      },
      {
        test: /\.less$/i,
        use: [
          MiniCssExtractPlugin.loader,
          "css-loader",
          {
            loader: "less-loader",
            options: {
              lessOptions: {
                javascriptEnabled: true,
              },
            },
          },
        ],
      },
      {
        test: /\.(woff|woff2)(\?v=\d+\.\d+\.\d+)?$/,
        // type: "asset/inline",
        type: "asset/resource",
        generator: {
          filename: "font/[hash][ext][query]",
        },
      },
      {
        test: /\.(jp(e*)g|svg)$/,
        type: "asset/resource",
        generator: {
          filename: "img/[hash][ext][query]",
        },
      },
      {
        test: /\.md$/,
        use: [
          {
            loader: "html-loader",
          },
          {
            loader: "markdown-loader",
          },
        ],
      },
    ],
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: "css/[name].css",
    }),
    new ProgressBarPlugin({ percentBy: "entries" }),
  ],
  // https://webpack.js.org/configuration/resolve/#resolvealias
  resolve: {
    alias: {
      components: path.resolve(dirPath, "../src/components/index.js"),
      containers: path.resolve(dirPath, "../src/containers/index.js"),
      providers: path.resolve(dirPath, "../src/providers/index.js"),
      hooks: path.resolve(dirPath, "../src/hooks/index.js"),
      utils: path.resolve(dirPath, "../src/utils/index.js"),
    },
  },
});

const prodConfig = (dirPath) => ({
  devtool: false,
  mode: "production",
  output: {
    publicPath: "assets/",
  },
  optimization: {
    minimize: true,
    minimizer: [
      new TerserPlugin({
        parallel: 4,
      }),
      new CssMinimizerPlugin({ parallel: 4, sourceMap: false }),
    ],
  },
  plugins: [
    new FaviconsWebpackPlugin({
      logo: "./src/assets/img/favicon.png",
      publicPath: "/assets/",
      prefix: "",
      inject: true,
      favicons: {
        icons: {
          appleStartup: false,
          coast: false,
          firefox: false,
          windows: false,
          yandex: false,
        },
      },
    }),
    new CleanWebpackPlugin({
      verbose: true,
      cleanOnceBeforeBuildPatterns: [
        path.join(dirPath, "../www/assets/**/*"),
        path.join(dirPath, "../www/worker.js"),
        path.join(dirPath, "../www/workbox-*"),
      ],
    }),
    new HtmlWebpackPlugin({
      template: "src/index.html",
      filename: "../index.html",
      showErrors: true,
      chunksSortMode: "auto",
      minify: {
        removeComments: false,
        collapseWhitespace: true,
        useShortDoctype: true,
        minifyJS: true,
        minifyCSS: true,
        minifyURLs: true,
      },
    }),
    new WorkboxPlugin.GenerateSW({
      swDest: "../worker.js",
      clientsClaim: true,
      skipWaiting: true,
      // modifyURLPrefix: {
      //   "assets/": "",
      // },
    }),
    new Dotenv({
      path: path.join(dirPath, `/.env.prod`),
      safe: true,
      systemvars: true,
    }),
    new BundleAnalyzerPlugin({
      analyzerMode: "static",
      openAnalyzer: false,
      reportFilename: "../analyzer.html",
    }),
  ],
});

const devConfig = (dirPath) => ({
  mode: "development",
  output: {
    publicPath: "/",
  },
  plugins: [
    new FaviconsWebpackPlugin({
      logo: "./src/assets/img/favicon.png",
      publicPath: "/",
      prefix: "",
      inject: true,
      favicons: {
        icons: {
          appleStartup: false,
          coast: false,
          firefox: false,
          windows: false,
          yandex: false,
        },
      },
    }),
    new HtmlWebpackPlugin({
      template: "src/index.html",
      filename: "index.html",
      showErrors: true,
      chunksSortMode: "auto",
      minify: {
        removeComments: false,
        collapseWhitespace: true,
        useShortDoctype: true,
        minifyJS: true,
        minifyCSS: true,
        minifyURLs: true,
      },
    }),
    new Dotenv({
      path: path.join(dirPath, `/.env.dev`),
      safe: true,
      systemvars: true,
    }),
  ],
  devServer: {
    contentBase: path.join(dirPath, "../../../www/"),
    index: "index.html",
    compress: true,
    open: true,
    port: 9090,
    hot: true,
    historyApiFallback: true,
  },
});

module.exports = (options = {}, dirPath) => (env) =>
  merge(
    baseConfig(dirPath),
    env.dev ? devConfig(dirPath) : prodConfig(dirPath),
    options
  );
